### ISSUE TITLE
As an editor, I want upload and manage page attachments in wagtail so that I can associate them with pages.

### ISSUE THREAD

**Issue by thatbudakguy** [Princeton-CDH/cdh-web#245]
# testing notes
in cdhweb v2, you could add "attachments" to most types of pages. an "attachment" could be either a document uploaded to the website (like a .pdf or a word document), or it could be a link that pointed to something elsewhere (like a google drive URL or a DOI for a dataset).

in cdhweb v3, we've tried to replicate this same experience, although the two kinds of attachment (local vs. external) are represented somewhat differently. for uploaded attachments, wagtail thinks of these natively as "Documents", and you can see them when you're on the main menu of wagtail. you can upload/edit Documents on their own, without needing to attach them to pages. for external attachments (links), we created a new category of Snippet, so they live in the same place as Page Intros.

to test this issue, try making both a local attachment (Document) and an external attachment (link). You can upload files and/or use any URL you want. after you make at least one of each type, go edit any page (can be a regular content page or a project page, profile, etc). towards the bottom of the editor view, you should see a new "attachments" field where you can select your two new attachments, in any order. once you do this and view the page on the site, you should see the special attachment "cards" at the bottom of the page, and they should look visually similar to the "cards" on production, for example:
 
<img width="1052" alt="Screen Shot 2021-03-22 at 2 44 23 PM" src="https://user-images.githubusercontent.com/4924494/112041739-1ce4c800-8b1d-11eb-961b-f9495aa03d94.png">

in the pic above, you can see an external attachment (link) on the annual report page. external attachments will have an "external link" icon, while locally uploaded ones will have an icon matching the type of the document (pdf, word doc, etc). you can try different file types, and also try adding attachments to different kinds of pages (e.g. profiles, projects, events).

# dev notes
<s>this could be a regular django migration, since we don't need to do anything complicated with streamfields, hierarchy, etc.</s> see #246; we're doing this via exodus.

two new models, both with m2m to `Page`:
- `Attachment`, which extends `AbstractDocument` and represents an uploaded file
- `ExternalLink`, which is not a `Document` subtype and represents content hosted outside the CDH site.

make sure that you're not losing access to the `page` object in context if the context object is renamed, because the attachments should be tied to it (not the specific page subclass) to access in the template. you will need to create an `AttachmentMixin`. you may want to check the old mezzanine admin config to see where the old attachment chooser was used.

the `AttachmentMixin` should have two fields: one m2m with `Attachment` and one m2m with `ExternalLink`. it also needs a property accessor to merge the lists of these two objects so they can be output in one loop in a template. it's fine to privilege local content (`Attachment`) by listing them first. the mixin will set up two separate chooser interfaces for the `Attachment`s (`DocumentChooserBlock`) and `ExternalLink`s (something else).

check what happens if a document is inserted via a `StreamField`. is this desirable? maybe we just need a "documents" section at the bottom, so that the attachments are guaranteed to render using the current design.

**Comment by rlskoeser** [comment]
I started looking at this to see if it's as easy as we hope. A quick attempt at extending wagtail's abstract document and configuring our attachment to be the default document resulted in a non-obvious error in the wagtail admin documents list.

Our attachment model is _either_ a link or an uploaded file. But wagtail documents have to have a file. So, I'm thinking migration path could be:
- create new link attachment model, and make it available as a wagtail snippet
- write data migration to convert all link-only attachments to new link attachment model
- remove link-specific fields from attachment and extend abstract document per https://docs.wagtail.io/en/v2.8.1/advanced_topics/documents/custom_document_model.html#module-wagtail.documents

Hopefully we can configure an attachment chooser that would let you select links or documents so the display logic can be similar to what we have now.

**Comment by thatbudakguy** [comment]
two models, both with generic foreign keys:
- `Attachment`, which extends `AbstractDocument` and represents an uploaded file
- `ExternalLink`, which is not a `Document` subtype and represents content hosted outside the CDH site.

**Comment by manawinters** [comment]
I was able to successfully upload a new document and attach it to a page, which appeared as it should on the site. When I tried adding an external link, however, I only had the option of using one of the existing links. It wouldn't allow me to upload a new one.
<img width="913" alt="Screen Shot 2021-03-26 at 10 14 06 AM" src="https://user-images.githubusercontent.com/79113717/112644802-32f3d080-8e1c-11eb-8d7d-3247ff13a7da.png">
